blog/2016-04-20-futhark-and-pygame.rst

```diff
@@ -0,0 +1,250 @@
+---
+title: Creating Interactive Futhark GUIs
+author: Niels G. W. Serup
+description: Using PyGame and PyOpenCL to interact with Futhark programs in simple graphical user interfaces.
+---
+
+In the previous blog post, Troels showed you how to use Futhark's
+PyOpenCL backend to generate PNG files with Python and NumPy.  There
+was also a short mention of *interactive* examples built with PyGame.
+In this blog post we will demonstrate how to add interactive visual
+interfaces to simulations implemented in Futhark.
+
+`PyGame <http://www.pygame.org/>`_ is a Python library for the SDL 1.2
+graphics framework.  Although a Python 3 port does exist, for
+compatibility reasons we use the original Python 2 library.
+Library-savvy readers might point out that a more modern library like
+pyglet or PySDL2 might be more... modern, but we really only require
+few features, so good old PyGame is sufficient for our basic examples.
+
+We currently have four GUI examples:
+
+  + `Mandelbrot Explorer <https://github.com/HIPERFIT/futhark-benchmarks/tree/master/misc/mandelbrot-explorer>`_
+  + `Fluid Simulation GUI <https://github.com/HIPERFIT/futhark-benchmarks/tree/master/accelerate/fluid/gui>`_
+  + `Game of Life <https://github.com/HIPERFIT/futhark-benchmarks/tree/master/misc/life>`_
+  + `HotSpot 2D Heat Equation GUI <https://github.com/HIPERFIT/futhark-benchmarks/tree/master/rodinia/hotspot>`_
+
+It is pretty easy to create an interactive visualisation with Futhark,
+so expect this number to go up.  However, note that Futhark
+development does *not* target interactive visualisations -- we just
+want to show off what we have!
+
+
+A Simple Start: Game of Life
+----------------------------
+
+Let us start with Futhark's own Game of Life visualisation.  This is
+arguably the simplest of the four visualisations, as it is actually
+not interactive.  However, it shows how few lines of code are needed
+to integrate Futhark, NumPy, and PyGame.
+
+First, here is a video of me running the visualisation!
+
+.. raw:: html
+
+   <video src="/static/life-2016.04.20.webm"
+          poster="/static/life-2016.04.20-poster.jpg"
+          controls
+          width="480" height="270">
+     Sorry, your browser doesn't support embedded WebM videos, but
+     don't worry, you can <a
+     href="/static/life-2016.04.20.webm">download it</a> and watch it
+     with a compatible video player!
+   </video>
+   <p style="font-style: italic; margin-top: 0;">Direct link: <a href="/static/life-2016.04.20.webm">life-2016.04.20.webm</a></p>
+
+Secondly, here is how it works: `life-gui.py
+<https://github.com/HIPERFIT/futhark-benchmarks/blob/master/misc/life/life-gui.py>`_
+first imports our three Game of Life variants::
+
+  import life
+  import quadlife
+  import quadlife_alt
+
```

Comment:
These Python modules are generated from Futhark programs by means of the `futhark-pyopencl` compiler.